### ISSUE TITLE
Implement gamma optimization option from TruePNG

### ISSUE THREAD

**Issue by shssoichiro** [shssoichiro/oxipng#77]
https://css-ig.net/articles/truepng.php

> in most of cases, it is better to disable any chunk that correct the way the image is displayed, specially for Web. gAMA chunk is not an exception, and the -g0 should delete it. However, TruePNG let you to merge the way the image is displayed instead of using gAMA chunk.
>
> be warned, it is a no back process, so you should make a backup of your file (or use -out switch). gAMA chunk can be added or deleted, but in this case, it changes RGB values in the new file.

**Comment by kornelski** [comment]
The correct solution for all of gAMA, cHRM, and iCCP is to convert to sRGB using color profile transformation. If you're OK having dependency on LCMS2, I can supply required code.

**Comment by shssoichiro** [comment]
Thanks! I don't mind adding that dependency.

**Comment by ace-dent** [comment]
Please *don't* make this the default behaviour.
Although the `gama` chunk is increasingly supported, it is not always correct.

**Comment by AlexTMjugador** [comment]
> Although the `gama` chunk is increasingly supported, it is not always correct.

Would you mind elaborating on what kind of errors you've seen with this chunk, @ace-dent?

**Comment by ace-dent** [comment]
@AlexTMjugador over the history of png, some art programs wrote this value incorrectly. Even when it was 'correct', it led to unexpected rendering differences (e.g. Mac vs PC authoring). Also, until quite recently, there was inconsistent rendering across platforms / applications. It was safer to strip any `gAMA` and assume that *all* png RGB data was encoded with intent of sRGB.  I appreciate the situation has changed... but for 'mastering' high quality images with deliberate color profiles, is it common to worry about optimisation in general? 🤔 
I think typically arbitrary colour info gets baked in by the editing software, and Users may be unaware...

**Comment by AlexTMjugador** [comment]
I've just stumbled upon this pure Rust library for ICC profile conversion: https://github.com/awxkee/moxcms

It seems pretty straightforward to use as a library, so I think it could be great to give it a try within Oxipng to normalize color spaces in images.

**Comment by andrews05** [comment]
Hm, I've been reluctant to actually pursue anything here. Colour transformations are lossy, strictly speaking, and will have an unknown impact on compressibility. E.g. it's possible (though less likely) that converting an iCCP to sRGB could actually come out larger.

That said, I have considered an experimental option to convert to XYB which generally compresses better.

**Comment by TPS** [comment]
> That said, I have considered an experimental option to convert to XYB which generally compresses better.

Forgive me if this is an answer that can simply Googled or LLMed, but practical colorspace stuff confuses me:

@andrews05 Is conversion to XYB lossless over (typically) _any_ colorspace that PNG is designed to handle?  Is such supported by decoders? _(I myself have only some recognition of XYB d/t its being handled natively in JxL, & so have some inkling it could be helpful.)_

**Comment by kornelski** [comment]
Generally you can't convert 8-bit sRGB (I mean 8 bits per channel, 24/32-bit per pixel) to anything else in 8 bits without some loss of precision, since images are already utilizing the maximum range (apart from edge cases like content that is already posterized or desaturated). 

You could convert transformed 8-bit into 16-bit depth, but then I'm sceptical whether doubling of the raw data size is going to be worth it (it decreases zlib's effective compression window), and whether it's actually going to be lessless, because decoders are going to have to have precise color management pipeline to accurately reverse it.

However, it may still be worth doing transformations of 8-bit to the slightly lossy 8-bit. A lot of content can survive a loss of one bit or even two, and that may be a good trade-off if it significantly reduces the file size.

**Comment by TPS** [comment]
> However, it may still be worth doing transformations of 8-bit to the slightly lossy 8-bit. A lot of content can survive a loss of one bit or even two, and that may be a good trade-off if it significantly reduces the file size.

@kornelski We recently had a [conversation](https://github.com/oxipng/oxipng/issues/689) where the consensus was to leave such to your excellent `pngquant` & then run `oxipng` thereafter.  Or do you mean something beyond that?

**Comment by andrews05** [comment]
> [@andrews05](https://github.com/andrews05) Is conversion to XYB lossless over (typically) _any_ colorspace that PNG is designed to handle? Is such supported by decoders? _(I myself have only some recognition of XYB d/t its being handled natively in JxL, & so have some inkling it could be helpful.)_

PNG itself doesn't really deal with colourspaces. It allows colourspaces to be specified via embedded ICC profiles but these are typically handled at the application level rather than by the decoder.
Generally only RGB-type colourspaces are suitable for use with PNG, and XYB is _not_ an RGB-type colourspace, so the same caveats would apply here as with XYB's use in jpegs by `jpegli`. Namely, if the application doesn't support ICC profiles, it will render horribly. So a hypothetical XYB option in oxipng would be generally not recommended, but could be interesting as an experiment. It might be okay for web usage, for instance, since all browsers have the necessary ICC support.

> [@kornelski](https://github.com/kornelski) We recently had a [conversation](https://github.com/oxipng/oxipng/issues/689) where the consensus was to leave such to your excellent `pngquant` & then run `oxipng` thereafter. Or do you mean something beyond that?

8-bit here is referring to 8-bit RGB (full colour), rather than 8-bit indexed which is what pngquant does.

**Comment by kornelski** [comment]
Sorry I wasn't clear. I've meant 8-bit *per channel* RGB/RGBA images, which means 24 or 32-bit pixels with 16.7M colors. In that case XYB or other tricks in 24/32-bit per pixel images would still give you millions of colors, but some lower number of millions. 

Conversion to 256-color palette is much more lossy than that, but also likely to give the smallest file sizes by far. Palette PNGs won't benefit further from any color transformations like XYB or gamma, because all of their color data takes a fixed 768 bytes.

The slightly-lossy color transformation tricks would be for images that have too many colors for a palette, but don't need to be exactly perfect. I expect it would be something like 10% file size saving that almost always looks lossless to they naked eye. 

Buuuut if you're going down the path of making PNG lossy, there are tricks for applying filters in a lossy way that can reduce files by 30-50% https://github.com/richgel999/rdopng/wiki/Examples and potentially lossy zlib implementations that could work similarly.

**Comment by TPS** [comment]
By similar logic to https://github.com/oxipng/oxipng/issues/689#issuecomment-2719450611, is it worth starting to do actual lossy transforms in `oxipng` when other tools already specialize in that & they could be improved by contributions there?  I suppose a kind of `rdopngquant` mashup + `fileoptimizer` might be the ultimate in that realm.

**Comment by andrews05** [comment]
There's also some brief discussions of lossy programs in #76.

@kornelski Is your offer still open to implement this? 😉

**Comment by kornelski** [comment]
There were multiple "this"es discussed here. So:

1. Application of color profiles (`iCCP` chunk) to the pixels to convert them to a profile-less sRGB if possible — yes. I think it's worth doing, and I can implement it. It can be a major file size saving, while also making files more reliably supported everywhere.  
   It's *slightly* lossy due to inevitable rounding and loss of precision, but it's "visually lossless", and proper decoders will probably do the same thing anyway, so I think it's fine to be in `oxipng`. It might be an opt-in CLI flag just to be careful.

2. Conversion of non-standard `gAMA` and `cHRM` to sRGB by applying them to the pixels could be done too. It's just a couple lines of code once there's color profile support. I could implement that too, but I'd consider that a low priority. This one isn't as certain to give file size savings, since these chunks are small.

3. Addition of new artificial `gAMA` chunks and transformations to the image data designed to make the pixels more compressible — no. I don't think it's worth doing.

4. Lossy compression by use of lossy filters, intentional posterization, or lossy gzip — no. These are clever techniques, and I'd love to implement them someday, but I don't have time for this currently. Such significant image quality changes are also better left to other tools than `oxipng`.

5. One more that hasn't been discussed, but could be considered: conversion of arbitrary `iCCP` files to equivalent smaller `iCCP` files, or just `gAMA` and `cHRM` chunks. Basically ICC files themselves could have an optimizer. This could be lossless (in the sense that differences between different ways of defining the profiles could be so small that when applied, they still round to the same 8-bit values). I've already implemented this hack for the special case of sRGB ICC files to `sRGB` chunk, but it could be generalized. I don't know if I'll have time to implement that.

**Comment by andrews05** [comment]
Right, yeah I'm really just thinking of #1. Although I've been reluctant to spend time on it myself (especially as I don't have much experience here), I do agree there could be value in an opt-in flag to do this.
